README.md
# danger-plugin-prfect

[![npm version][npm-img]][npm-url]

> Enforce guidelines for high-quality pull requests.

```sh
npm install --save-dev danger-plugin-prfect
```

## Usage

```js
// dangerfile.js
import { checkTitle } from 'danger-plugin-prfect'

checkTitle()
```

### checkTitle

Checks for subject length, ~~capitalisation,~~ ~~trailing punctuation,~~ and
imperative mood.

[npm-url]: https://www.npmjs.com/package/danger-plugin-prfect
[npm-img]: https://img.shields.io/npm/v/danger-plugin-prfect.svg?style=flat-square


CHANGELOG.md
# danger-plugin-prfect

### 0.1.1 — 2022-02-24

#### 🐛 Fixed

- Fix type declaration output [`03a9ad9`](https://github.com/stormwarning/danger-plugin-prfect/commit/03a9ad9595cf1175d6ff561d15ac31c583d57b06)
- Remove hard line break in `LENGTH_TIP` [`2ef35a`](https://github.com/stormwarning/danger-plugin-prfect/commit/2ef35a3fd2e88d9b947ec6927636400426ea63a2)

## 0.1.0 — 2022-02-18

#### 🎁 Added

- Export plugin function types [#1](https://github.com/stormwarning/danger-plugin-prfect/pull/1)
- Add `checkTitle` output options [#3](https://github.com/stormwarning/danger-plugin-prfect/pull/3)

  Allows for customising response level. Also returns an object of results
  to allow consumers to craft their own response message.


src/index.test.ts
// Disable tests for now
// import prfect from './index'

// declare const global: any

// describe('prfect()', () => {
// 	beforeEach(() => {
// 		global.warn = jest.fn()
// 		global.message = jest.fn()
// 		global.fail = jest.fn()
// 		global.markdown = jest.fn()
// 	})

// 	afterEach(() => {
// 		global.warn = undefined
// 		global.message = undefined
// 		global.fail = undefined
// 		global.markdown = undefined
// 	})

// 	it('Checks for a that message has been called', () => {
// 		global.danger = {
// 			github: { pr: { title: 'My Test Title' } },
// 		}

// 		prfect()

// 		expect(global.message).toHaveBeenCalledWith('PR Title: My Test Title')
// 	})
// })
// eslint-disable-next-line unicorn/no-empty-file


src/__tests__/index.test.ts
import anyTest, { TestFn } from 'ava'

import { checkTitle } from '../../dist/index.js'
// eslint-disable-next-line import/extensions
import { IMPERATIVE_ERROR, LENGTH_ERROR, LENGTH_WARNING } from '../messages'

const test = anyTest as TestFn<GlobalContext>

interface GlobalContext {
	warn: string[]
	message: string[]
	fail: string[]
	markdown: string[]
}

declare const global: {
	warn: (m: string) => void
	message: (m: string) => void
	fail: (m: string) => void
	markdown: (m: string) => void
	danger: unknown
}

test.beforeEach((t) => {
	t.context.warn = []
	t.context.message = []
	t.context.fail = []
	t.context.markdown = []
	global.warn = (m: string) => t.context.warn.push(m)
	global.message = (m: string) => t.context.message.push(m)
	global.fail = (m: string) => t.context.fail.push(m)
	global.markdown = (m: string) => t.context.markdown.push(m)
})

function checkPr({ title }: { title: string }) {
	global.danger = {
		github: { pr: { title } },
	}

	checkTitle()
}

test.serial('show failure when not using imperative mood', async (t) => {
	checkPr({ title: 'Fixes a bug in production' })
	t.true(t.context.fail.includes(IMPERATIVE_ERROR))
})

test.serial('shows no message when using imperative mood', async (t) => {
	checkPr({ title: 'Fix a bug in production' })
	t.false(t.context.fail.includes(IMPERATIVE_ERROR))
})

test.serial('can set a custom response level', async (t) => {
	checkTitle('fixes a bug in production.', { imperativeMood: 'warn' })
	t.false(t.context.fail.includes(IMPERATIVE_ERROR))
	t.true(t.context.warn.includes(IMPERATIVE_ERROR))
})

test.serial('warns when a title is between 50 and 72 characters', async (t) => {
	checkTitle('warns when a title is between 50 and 72 characters.')
	t.true(t.context.warn.includes(LENGTH_WARNING))
})

test.serial(
	'shows a failure when a title is greater than 72 characters',
	async (t) => {
		checkTitle(
			'Advanced bullshit cutting-edge cable Chiba sanpaku amalgam unfolding psychoactive program.'
		)
		t.true(t.context.fail.includes(LENGTH_ERROR))
	}
)

test.serial('logs no message when set to returnOnly', async (t) => {
	checkTitle('fixes a bug in production.', { returnOnly: true })
	t.is(t.context.fail.length, 0)
	t.is(t.context.warn.length, 0)
	t.is(t.context.message.length, 0)
	t.is(t.context.markdown.length, 0)
})


zazen.config.js
// {
//     "compilerOptions": {
//         "target": "es6",
//         "module": "commonjs",
//         "rootDir": "src",
//         "outDir": "dist",
//         "allowJs": false,
//         "pretty": true,
//         "strictNullChecks": true,
//         "declaration": true
//     },
//     "lib":["es2017"],
//     "include": [
//         "src/**/*.ts",
//         "src/**/*.tsx",
//         "dangerfile.ts"
//     ],
//     "exclude": [
//         "dangerfile.ts",
//         "node_modules",
//         "dist"
//     ]
// }


export default {
	setEslintConfig: (config) => ({
		...config,
		extends: [
			...(config.extends),
			'@zazen/eslint-config/node',
			'@zazen/eslint-config/typescript',
		],
		rules: {
			...config.rules,
			'yoda': 'off',
			'@typescript-eslint/no-require-imports': 'off',
			'@typescript-eslint/no-unsafe-return': 'off',
			'import/no-anonymous-default-export': [
				'error',
				{ allowObject: true },
			],
		},
	}),
}


src/index.ts
import nlp from 'compromise'
// Provides dev-time type structures for  `danger` - doesn't affect runtime.
import { DangerDSLType } from 'danger/distribution/dsl/DangerDSL.js'

import {
	IMPERATIVE_ERROR,
	IMPERATIVE_TIP,
	LENGTH_ERROR,
	LENGTH_TIP,
	LENGTH_WARNING,
} from './messages.js'

declare const danger: DangerDSLType
// Declare function message(message: string): void
declare function warn(message: string): void
declare function fail(message: string): void
declare function markdown(message: string): void

interface CheckTitleOptions {
	/**
	 * Error type for imperative mood check.
	 * @default 'fail'
	 */
	imperativeMood?: false | 'warn' | 'fail'
	/**
	 * Should a character limit be enforced.
	 * @default true
	 */
	titleLength?: boolean
	returnOnly?: boolean
}

type CheckResult = 'pass' | 'warn' | 'fail'

interface CheckOutput {
	result?: CheckResult
	message?: string
	tip?: string
}

/**
 * Validate PR titles
 *
 * Checks for subject length, capitalisation, trailing punctuation,
 * and imperative mood.
 */
export function checkTitle(title?: string, options?: CheckTitleOptions) {
	title ||= danger.github.pr.title
	let { imperativeMood, titleLength, returnOnly } = options || {
		imperativeMood: 'fail',
		titleLength: true,
		returnOnly: false,
	}

	let imperativeMoodOutput: CheckOutput
	if (imperativeMood) {
		/** @todo Ignore "tag" prefixes. */
		let isImperative = Boolean(nlp(title).verbs().isImperative().text())

		if (!isImperative && !returnOnly) {
			if (imperativeMood === 'fail') fail(IMPERATIVE_ERROR)
			if (imperativeMood === 'warn') warn(IMPERATIVE_ERROR)
			markdown(IMPERATIVE_TIP)
		}

		imperativeMoodOutput = {
			result: isImperative ? 'pass' : imperativeMood,
			message: !isImperative && IMPERATIVE_ERROR,
			tip: !isImperative && IMPERATIVE_TIP,
		}
	}

	let titleLengthOutput: CheckOutput
	if (titleLength) {
		let lengthResult: CheckResult = 'pass'
		let lengthMessage: string
		if (50 <= title.length && title.length <= 72) {
			if (!returnOnly) warn(LENGTH_WARNING)
			lengthResult = 'warn'
			lengthMessage = LENGTH_WARNING
		}

		if (72 < title.length) {
			if (!returnOnly) fail(LENGTH_ERROR)
			lengthResult = 'fail'
			lengthMessage = LENGTH_ERROR
		}

		let showLengthTip = false
		if (50 <= title.length) {
			if (!returnOnly) markdown(LENGTH_TIP)
			showLengthTip = true
		}

		titleLengthOutput = {
			result: lengthResult,
			message: lengthMessage,
			tip: showLengthTip && LENGTH_TIP,
		}
	}

	return {
		imperativeMood: imperativeMoodOutput,
		titleLength: titleLengthOutput,
	}
}
